### PR TITLE
Add FocusWindow options to select by index, or filter by passed func

### PR DESCRIPTION
**Pull request by daanzu**
Has been helpful for me for a long time. Only recently integrated into upstream.

### CODE REVIEW

**Comment by Danesprite** [comment]
Thanks for putting this together. I can see it being very useful!

File: dragonfly/actions/action_focuswindow.py [hunk 1]
```diff
@@ -40,34 +40,54 @@ class FocusWindow(ActionBase):
...
     """
 
-    def __init__(self, executable=None, title=None):
+    def __init__(self, executable=None, title=None, index=None,
+            filter_func=None):
         if executable:  self.executable = executable.lower()
         else:           self.executable = None
         if title:       self.title = title.lower()
         else:           self.title = None
+        if index:       self.index = index.lower()
```
Comment (Danesprite) [inline dragonfly/actions/action_focuswindow.py]: Using `lower()` here is making `int` arguments fail

File: dragonfly/actions/action_focuswindow.py [hunk 2]
```diff
@@ -40,34 +40,54 @@ class FocusWindow(ActionBase):
...
     def _execute(self, data=None):
         executable = self.executable
         title = self.title
+        index = self.index
         if data and isinstance(data, dict):
             if executable:  executable = (executable % data).lower()
             if title:       title = (title % data).lower()
+            if index:       index = (index % data).lower()
+
+        index = int(index) if index else 0
```
Comment (Danesprite) [inline dragonfly/actions/action_focuswindow.py]: I didn't know this worked without raising a `ValueError`. Neat!

**Comment by Danesprite** [comment]
Thinking about it, I think these two extra options should also be added to `BringApp` and passed to the internal `FocusWindow` action it uses.

**Comment by daanzu** [comment]
Makes sense. I've never used `BringApp` myself, so I forget it's there.

**Comment by Danesprite** [comment]
Yeah, I don't use it much myself either. I'll add those changes in another PR shortly.

**Pull request merged**